File: Frontend/angular/src/api/services/product.services.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';
import { PaginationType } from '../models/pagination';
import { Product } from '../models/product';

type PageReturn = {
  results: Product[];
  pagination: PaginationType;
};
@Injectable()
export class ProductService {
  private readonly PATH: string = '/product';

  constructor(private http: HttpClient) {}

  getProductPage(
    page = 0,
    size = 25,
    search = '',
    order = 'asc',
    sort = 'name'
  ): Observable<PageReturn> {
    return this.http.get<PageReturn>(
      environment.baseUrl +
        this.PATH +
        `/pages?page=${page}&size=${size}&search=${search}&order=${order}&sort=${sort}`
    );
  }
  getProductById(id: number): Observable<Product> {
    return this.http.get<Product>(environment.baseUrl + this.PATH + `/${id}`);
  }
  insert(product: Product): Observable<Product> {
    return this.http.post<Product>(environment.baseUrl + this.PATH, product);
  }
  update(product: Product): Observable<Product> {
    return this.http.patch<Product>(
      environment.baseUrl + this.PATH + '/' + product.id,
      product
    );
  }
  remove(id: number): Observable<void> {
    return this.http.delete<void>(environment.baseUrl + this.PATH + `/${id}`);
  }
}

File: Mobile/flutter_scv/lib/app/services/product.services.dart
import 'package:dio/dio.dart';
import 'package:flutter_scv/app/models/pagination.dart';
import 'package:flutter_scv/app/models/product.dart';
import 'package:flutter_scv/environments.dart';

class ProductServices {
  final endPoint = '/product';
  Dio dio = Dio();

  ProductServices() {
    dio.options.connectTimeout = 15000;
    dio.options.receiveTimeout = 15000;
    dio.options.sendTimeout = 15000;
    dio.options.baseUrl = EnvironmentConfig.urlsConfig();
  }

  Future getProductPage({required page, search = ''}) async {
    List listProduct;

    try {
      final response =
          await dio.get('$endPoint/pages?page=$page&search=$search');
      listProduct =
          response.data['results'].map((item) => Product.fromJson(item)).toList();
      Pagination pagination = Pagination(
          length: response.data['pagination']['length'],
          size: response.data['pagination']['size'],
          lastPage: response.data['pagination']['lastPage'],
          page: response.data['pagination']['page'],
          startIndex: response.data['pagination']['startIndex'],
          endIndex: response.data['pagination']['endIndex']);

      return [listProduct,pagination];
    } on DioError catch (e) {
      return Future.error(e.response!.data["message"]);
    }
  }

  Future<Product> getProductById(int id) async {
    try {
      final response = await dio.get('$endPoint/$id');
      return Product.fromJson(response.data);
    } on DioError catch (e) {
      return Future.error(e.response!.data["message"]);
    }
  }

  Future insert(Product product) async{
    try{
      final response = await dio.post(endPoint,data:{product});
      return response.data['id'];
    }on DioError catch (e) {
      return Future.error(e.response!.data["message"]);
    }
  }

  Future update(Product product) async{
    try{
      final response = await dio.patch('$endPoint/${product.id}',data:product);
    }on DioError catch (e) {
      return Future.error(e.response!.data["message"]);
    }
  }

  Future remove(int id) async{
    try{
      final response = await dio.delete('$endPoint/$id');
    }on DioError catch (e) {
      return Future.error(e.response!.data["message"]);
    }
  }
}
